c10/util/intrusive_ptr.h

```diff
@@ -207,16 +225,37 @@ class intrusive_ptr final {
     target_ = NullType::singleton();
   }
 
+  // raw pointer constructors are not public because we shouldn't make
+  // intrusive_ptr out of raw pointers except from inside the make_intrusive(),
+  // reclaim() and weak_intrusive_ptr::lock() implementations.
+
   // This constructor will not increase the ref counter for you.
-  // This is not public because we shouldn't make intrusive_ptr out of raw
-  // pointers except from inside the make_intrusive() and
-  // weak_intrusive_ptr::lock() implementations
-  explicit intrusive_ptr(TTarget* target) noexcept : target_(target) {}
+  // We use the tagged dispatch mechanism to explicitly mark this constructor
+  // to not increase the refcount
+  explicit intrusive_ptr(TTarget* target, raw::DontIncreaseRefcount) noexcept
+      : target_(target) {}
+
+  // This constructor will increase the ref counter for you.
+  // This constructor will be used by the make_intrusive(), and also pybind11,
+  // which wrap the intrusive_ptr holder around the raw pointer and incref
+  // correspondingly (pybind11 requires raw pointer constructor to incref by
+  // default).
+  explicit intrusive_ptr(TTarget* target)
+      : intrusive_ptr(target, raw::DontIncreaseRefcount{}) {
+    if (target_ != NullType::singleton()) {
+      // We can't use retain_(), because we also have to increase weakcount
+      // and because we allow raising these values from 0, which retain_()
+      // has an assertion against.
+      ++target_->refcount_;
+      ++target_->weakcount_;
```

Comment:
Doesn't this introduce the possibility of a memory leak? If `target` already has one intrusive_ptr referencing it, this will set weakcount_ to 2, meaning that ~intrusive_ptr will never be able to `delete target`.

Reply:
while it's possible, but this is a private constructor and it's not intended to use by users who are trying to create a intrusive_ptr from T*, the only way user can create intrusive_ptr is from `make_intrusive`, which is guaranteed that there's no refcount already referring to it.